### ISSUE TITLE
Can't change 27753 port for 'iOSDeviceManager' /  DeviceAgent-Runner.app

### ISSUE THREAD

**Issue by Goginenni** [calabash/run_loop#541]
Hi,
I tried according to this post https://github.com/calabash/run_loop/issues/540 but doesn't work for me.

DEVICE_AGENT_URL=http://127.0.0.1:8086 ./run_loop/2.2.0/lib/run_loop/device_agent/bin/iOSDeviceManager start_test --device-id 026B9D52-492B-4FB1-B8C6-B48B83A27E06

$ curl -X GET '-H "Content-Type: application/json"' http://127.0.0.1:8086/1.0/health
curl: (7) Failed to connect to 127.0.0.1 port 8086: Connection refused

But

$ curl -X GET '-H "Content-Type: application/json"' http://127.0.0.1:27753/1.0/health
{"status":"Calabash is ready and waiting."}

How can i change this port? 

Regards,
Sathish

**Comment by jmoody** [comment]
Override the `:port` option.

https://github.com/calabash/run_loop/blob/develop/lib/run_loop/device_agent/client.rb#L33

**Comment by Goginenni** [comment]
@jmoody Are you sure?. For me, Overriding this port only changes URL for Device agent client to connect. But it is not changing the actual port in the DeviceAgent-Runner.app.

**Comment by jmoody** [comment]
That is correct.  My mistake.  There is no way to change this port number on the DeviceAgent itself.

**Comment by Goginenni** [comment]
@jmoody Is it apple limitation ? or it a hardwired value that could in principle be changed.

**Comment by jmoody** [comment]
It could be changed but there is no mechanism for doing so.  What is your use case?

**Comment by Goginenni** [comment]
I want to run the tests parallel in iOS simulators and this issue is blocking us.

**Comment by Goginenni** [comment]
@jmoody Thanks.

**Comment by jmoody** [comment]
@Goginenni This is not on the critical path at the moment.  I have reopened this issue and put it into the backlog.

**Comment by Goginenni** [comment]
@jmoody Ok. it is quite critical to us and I would like to  dig into it & fix. Can i see iOSDeviceManager code or it is impossible to share this code ?

**Comment by MortenGregersen** [comment]
+1
This is also critical for us. We really need to be able to run on multiple simulators at once, to get our test executed in a minimum amount of time.

**Comment by maxelrod** [comment]
+1 This too is something we would like.

**Comment by astapinski** [comment]
+1 did anything ever come of this? I'd really like to be able to run on multiple simulators.

**Comment by RadiusGarrett** [comment]
+1 my company runs hundreds of calabash tests and we wish to gate quality on the PR level, but without launching multiple simulators turnaround time is too long with the number of tests needed to run. Can this be prioritized?

**Comment by RadiusGarrett** [comment]
@jmoody 
+1 my company runs hundreds of calabash tests and we wish to gate quality at the PR level to prevent regression, but without launching multiple simulators turnaround time is too long with the number of tests needed to run.

What can the public do to help prioritize this feature?

**Comment by jmoody** [comment]
Sadly, the public can do nothing until DeviceAgent.iOS open sourced.

If you are Xamarin Test Cloud or Microsoft Mobile Center customers, you can request this feature from your representative.

To temper your expectations, this feature is an extremely low priority for Microsoft.

**Comment by RadiusGarrett** [comment]
Understood and thank you for the prompt reply.

**Comment by jmoody** [comment]
I have been thinking about running multiple simulators concurrently.  I don't think this is something we can easily support.  The port issue is somewhat trivial - I can think of a solution.  The real problem is managing the underlying `xcodebuild test-without-building` processes.  There are several scenarios where an existing `xcodebuild` process will reopen the Simulator after a test is terminated.   This causes the next test to fail.  My solution is to enforce a "one xcodebuild process targeting a Simulator" rule.

We switched to `xcodebuild test-without-building` because the FBSimulatorControl start-test reverse engineering solution stopped working in Xcode 8.3.3.  We followed their lead and switched to calling xcodebuild.  The community could investigate whether or not the previous FBSimulatorControl start-test solution can be revived.

We could also try transitioning to using FBSimulatorControl's fbsimctl to manage the simulators - I don't know if that tool also has this relaunch problem.

**Comment by RadiusGarrett** [comment]
@jmoody 

Will the ability to launch multiple concurrent simulators in Xcode 9 help at all?

https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/WhatsNewXcode/xcode_9/xcode_9.html

"Simulator
- New in Xcode 9 - Multiple concurrent simulators.
- Run multiple simulators at the same time.
- Run more tests in parallel.
- Test synching and other multi-device workflows."

**Comment by jmoody** [comment]
> Will the ability to launch multiple concurrent simulators in Xcode 9 help at all?

No.  This is not related to the unchangeable port issue.

**Comment by blamay** [comment]
@jmoody 

Do you see any reason why a sed/perl command to change the ports wouldn't work? 

Something like this https://stackoverflow.com/questions/15402770/how-to-grep-and-replace

**Comment by jmoody** [comment]
I will repeat:

Sadly, the public can do nothing until DeviceAgent.iOS open sourced.

If you are Xamarin Test Cloud or Microsoft Mobile Center customers, you can request this feature from your representative.

To temper your expectations, this feature is an extremely low priority for Microsoft.

The port number is _compiled into_ the DeviceAgent.app.

**Comment by blamay** [comment]
> The port number is compiled into the DeviceAgent.app

This answers my question. I wanted to be sure before investing in additional mac machines. Thank you for the reply.
